### PR TITLE
Fix bug of removing samlssoTokenId cookie when remember me option is selected

### PR DESCRIPTION
**Pull request by bhagyasakalanka**
## Related Issues
- https://github.com/wso2/product-is/issues/10934
- https://github.com/wso2-enterprise/wso2-iam-internal/issues/42

## Related PRs
- https://github.com/wso2-extensions/identity-inbound-auth-saml/pull/341

### CODE REVIEW

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java [hunk 1]
```diff
@@ -1396,15 +1397,24 @@ private Cookie getTokenIdCookie(HttpServletRequest req) {
...
-                                    String tenantDomain, String loggedInTenantDomain) {
+                                    String tenantDomain, String loggedInTenantDomain, String sessionIdentifier) {
 
         ServletCookie samlssoTokenIdCookie = new ServletCookie(SAML_SSO_TOKEN_ID_COOKIE, sessionId);
         IdentityCookieConfig samlssoTokenIdCookieConfig = IdentityUtil
                 .getIdentityCookieConfig(SAML_SSO_TOKEN_ID_COOKIE);
 
+        // Get age of the samlssoTokenId cookie.
+        SessionContext sessionContext = getSessionContext(sessionIdentifier, loggedInTenantDomain);
+        Integer cookieAge = null;
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java:1413]:
```suggestion
        Integer cookieAge = 0;
```

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java [hunk 2]
```diff
@@ -2234,4 +2246,18 @@ private void sendNotificationForECPRequest(HttpServletResponse resp, String erro
...
+    /**
+     * To retrieve session context of the given session identifier.
+     *
+     * @param sessionIdentifier Session identifier.
+     * @param loginTenantDomain Login tenant domain.
+     * @return Session context for the given session identifier.
+     */
+    private SessionContext getSessionContext(String sessionIdentifier, String loginTenantDomain) {
+
+        if (sessionIdentifier != null && loginTenantDomain != null) {
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java:2258]: use StringUtils and check for empty and null

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java [hunk 3]
```diff
@@ -1424,9 +1434,11 @@ private void storeTokenIdCookie(String sessionId, HttpServletRequest req, HttpSe
         }
 
         samlssoTokenIdCookie.setSameSite(SameSiteCookie.NONE);
-
+        if (cookieAge != null) {
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java:1437]: we don't need null check if we initialize the max age with 0

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java (hunk 1 above)
Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java:1413]: We cannot initialize cookieAge to 0. If we initialized to 0, then the cookie will not created as the age is expired. That's why this null check is introduced in https://github.com/wso2-extensions/identity-inbound-auth-saml/pull/341. Also we cannot check if the cookieAge == 0 and ignore as when removing the cookie, we set cookie value to 0 to make it expire.

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java (hunk 2 above)
Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java]: Fixed

**Review by piraveena — APPROVED** [review]
Run the PR builder

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java [hunk 4]
```diff
@@ -1396,15 +1397,24 @@ private Cookie getTokenIdCookie(HttpServletRequest req) {
      * @param sessionId            Session Id.
      * @param req                  HttpServlet Request.
      * @param resp                 HttpServlet Response.
+     * @param tenantDomain         Tenant Domain
      * @param loggedInTenantDomain Logged In Tenant Domain.
+     * @param sessionIdentifier    Session Identifier
      */
     private void storeTokenIdCookie(String sessionId, HttpServletRequest req, HttpServletResponse resp,
-                                    String tenantDomain, String loggedInTenantDomain) {
+                                    String tenantDomain, String loggedInTenantDomain, String sessionIdentifier) {
```
Comment (mevan-karu) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java:1405]: What is the difference of `sessionId` and `sessionIdentifier` params?

Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java:1405]: sessionId is the cookie value of samlssoTokenId cookie (Could be a random UUID as well). sessionIdentifier is used to uniquely identify the session. Those values are different from one another.

**Pull request merged**